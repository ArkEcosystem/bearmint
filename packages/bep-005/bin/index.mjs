#!/usr/bin/env node
import { join } from 'node:path'

import { globby } from 'globby'
import meow from 'meow'

process.env.NODE_ENV = 'test'

const { flags } = meow({
	flags: {
		config: {
			type: 'string',
		},
	},
	importMeta: import.meta,
})

const { files } = (await import(join(process.cwd(), flags.config))).default
const tests = (await globby(files)).map((path) => join(process.cwd(), path))

for (const test of tests) {
	await import(test)
}
