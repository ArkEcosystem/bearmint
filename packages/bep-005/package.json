{
	"name": "@bearmint/bep-005",
	"version": "0.0.0",
	"private": true,
	"description": "This library is the official reference implementation for BEP-005 (https://bep.dev/005)",
	"keywords": [
		"bearmint",
		"tendermint",
		"blockchain",
		"framework",
		"modular"
	],
	"homepage": "https://github.com/bearmint/bearmint",
	"repository": {
		"type": "git",
		"url": "git+https://github.com/bearmint/bearmint.git"
	},
	"funding": "https://funding.bearmint.com/",
	"license": "MIT",
	"author": "Bearmint Mining Company <hello@bearmint.com> (https://bearmint.com/)",
	"type": "module",
	"exports": "./distribution/index.js",
	"types": "./distribution/index.d.ts",
	"bin": {
		"roadrunner": "./bin/index.mjs"
	},
	"files": [
		"bin",
		"distribution"
	],
	"scripts": {
		"audit": "pnpm audit --fix --prod",
		"build": "pnpm run clean && tsc",
		"build:watch": "pnpm run clean && tsc -w",
		"clean": "rm -rf distribution",
		"depcheck": "depcheck --ignores=@bearmint/bep-004,@bearmint/bep-005,@bearmint/bep-031,@bearmint/bep-039,globby,meow",
		"deps:major": "ncu -u --target=latest",
		"deps:minor": "ncu -u --target=minor",
		"deps:patch": "ncu -u --target=patch"
	},
	"dependencies": {
		"@bearmint/bep-013": "workspace:^",
		"@bearmint/bep-018": "workspace:^",
		"@bearmint/bep-116": "workspace:^",
		"@sinonjs/fake-timers": "9.1.2",
		"ansi-styles": "6.2.1",
		"concordance": "5.0.4",
		"globby": "13.1.2",
		"lodash.get": "4.4.2",
		"lodash.has": "4.5.2",
		"meow": "^11.0.0",
		"nock": "13.2.9",
		"pretty-ms": "8.0.0",
		"sinon": "14.0.1",
		"strip-ansi": "7.0.1",
		"vitest": "^0.24.5"
	},
	"devDependencies": {
		"@types/node": "18.11.9",
		"@types/sinon": "10.0.13",
		"@types/sinonjs__fake-timers": "8.1.2"
	},
	"engines": {
		"node": ">=18",
		"pnpm": ">=7.9.0"
	},
	"publishConfig": {
		"access": "restricted"
	}
}
