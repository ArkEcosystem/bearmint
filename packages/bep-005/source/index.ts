export * from './expect.js'
export * from './fake.js'
export * from './nock.js'
export * from './spy.js'
export * from './stub.js'
