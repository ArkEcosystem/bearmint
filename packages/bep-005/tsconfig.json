{
	"compilerOptions": {
		"typeRoots": ["node_modules/@types"],
		"outDir": "distribution"
	},
	"exclude": ["node_modules", "test/**/**.ts"],
	"extends": "../../config/tsconfig.json",
	"include": ["source/**/**.ts"]
}
