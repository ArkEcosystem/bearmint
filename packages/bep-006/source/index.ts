export * as config from './config/index.js'
export * from './fixtures.js'
export * from './tendermint.js'
export * from './types.js'
