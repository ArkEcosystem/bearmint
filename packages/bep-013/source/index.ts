export * from './account.js'
export * from './app.js'
export * from './config.js'
export * from './consensus.js'
export * from './container.js'
export * from './crypto.js'
export * from './data.js'
export * from './event.js'
export * from './gas.js'
export * from './json.js'
export * from './logger.js'
export * from './milestone.js'
export * from './policies.js'
export * from './reward.js'
export * from './service-provider.js'
export * from './slash.js'
export * from './state.js'
export * from './store.js'
export * from './strategies.js'
export * from './tendermint.js'
export * from './tx.js'
export * from './utils.js'
export type { ClassOrFunctionReturning, Constructor, FunctionReturning } from 'awilix'

export const PROTO_OPTS = { readUnknownFields: false }
