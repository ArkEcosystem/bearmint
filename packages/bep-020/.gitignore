account
account_name
account_public_key
account_validator_address
application
milestone
tx
tx_receipt
tx_state
world
