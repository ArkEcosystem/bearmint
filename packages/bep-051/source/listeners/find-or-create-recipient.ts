import type { Cradle, EventListener } from '@bearmint/bep-013'
import { ErrorCode, TxAuditException } from '@bearmint/bep-109'

import { findOrCreateAccountByAddress } from '../utilities/find-or-create-account-by-address.js'
import type { TxListenerPayload } from './types.js'

export function makeFindOrCreateRecipientListener(
	cradle: Cradle,
): EventListener<TxListenerPayload> {
	return {
		async execute(data) {
			try {
				await findOrCreateAccountByAddress(
					data.state,
					cradle.AddressFactory,
					cradle.AccountSerializer,
					data.tx.data.recipient.toString(),
				)
			} catch {
				throw new TxAuditException({
					key: 'recipient',
					value: ErrorCode.TX_MALFORMED_ADDRESS,
				})
			}
		},
	}
}
