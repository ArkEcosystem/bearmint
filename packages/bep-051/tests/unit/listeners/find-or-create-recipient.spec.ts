import { expect, describe, it, beforeEach } from 'vitest'
import { extendExpect } from '@bearmint/bep-005'
extendExpect(expect)
import { milestone } from '@bearmint/bep-006'
import {
	fakeAccountSerializer,
	fakeAddressFactory,
	fakeCommittedState,
	fakeMultiStore,
} from '@bearmint/bep-008'
import type { EventRecorder, StateStore } from '@bearmint/bep-013'
import { ContainerType } from '@bearmint/bep-013'
import { makeContainer } from '@bearmint/bep-014'
import { makeEventDispatcher } from '@bearmint/bep-015'
import { makeState } from '@bearmint/bep-020'
import { makeAccountSerializer, makeAccountRepository } from '@bearmint/bep-022'
import { makeEventRecorder } from '@bearmint/bep-090'

import { makeFindOrCreateRecipientListener } from '../../../source/listeners/find-or-create-recipient.js'

describe<{
	eventRecorder: EventRecorder
	state: StateStore
}>('Find or Create Recipient', () => {
	beforeEach(async (context) => {
		const container = makeContainer()
		container.bindValue(ContainerType.DeliverTxState, fakeCommittedState())

		context.eventRecorder = makeEventRecorder()

		context.state = await makeState(
			makeAccountRepository({
				AccountSerializer: fakeAccountSerializer(),
				AddressFactory: fakeAddressFactory(),
				EventDispatcher: makeEventDispatcher(),
			}),
			await fakeMultiStore(),
		)

		context.state.setMilestone(milestone as any)
	})

	it('should capture an event if the tx recipient cannot be verified', async (context) => {
		context.eventRecorder.actingAs('find-or-create-recipient-fail-verify')

		await makeFindOrCreateRecipientListener({
			AccountSerializer: makeAccountSerializer({
				CommittedState: context.state,
				DeliverTxState: context.state,
				EventDispatcher: makeEventDispatcher(),
			}),
			AddressFactory: fakeAddressFactory(),
		}).execute({
			eventRecorder: context.eventRecorder,
			state: context.state,
			tx: {
				data: {
					recipient: {
						toString: () => 'bear1vlezq7dtyml3zu3wgfdwdglcalradfnt5xpzahl7a5478hn5g20qxd5nkr',
						verify: () => false,
					},
				},
			},
		})

		expect(context.eventRecorder.errors()).toMatchSnapshot()
	})

	it('should exit early if the recipient account already exists', async (context) => {
		context.eventRecorder.actingAs('find-or-create-recipient-already-exists')

		await context.state.getAccountRepository().index([
			await makeAccountSerializer({
				CommittedState: context.state,
				DeliverTxState: context.state,
				EventDispatcher: makeEventDispatcher(),
			}).deserialize({
				address: 'bear1vlezq7dtyml3zu3wgfdwdglcalradfnt5xpzahl7a5478hn5g20qxd5nkr',
			}),
		])

		await makeFindOrCreateRecipientListener({
			AccountSerializer: makeAccountSerializer({
				CommittedState: context.state,
				DeliverTxState: context.state,
				EventDispatcher: makeEventDispatcher(),
			}),
			AddressFactory: fakeAddressFactory(),
		}).execute({
			eventRecorder: context.eventRecorder,
			state: context.state,
			tx: {
				data: {
					recipient: {
						toString: () => 'bear1vlezq7dtyml3zu3wgfdwdglcalradfnt5xpzahl7a5478hn5g20qxd5nkr',
						verify: () => true,
					},
				},
			},
		})

		expect(context.eventRecorder.errors()).toMatchSnapshot()
	})

	it('should create a new account if the recipient account does not exist', async (context) => {
		context.eventRecorder.actingAs('find-or-create-recipient-does-not-exist')

		expect(
			await context.state
				.getAccountRepository()
				.hasByAddress('bear1vlezq7dtyml3zu3wgfdwdglcalradfnt5xpzahl7a5478hn5g20qxd5nkr'),
		).toBeFalse()

		await expect(
			makeFindOrCreateRecipientListener({
				AccountSerializer: makeAccountSerializer({
					CommittedState: context.state,
					DeliverTxState: context.state,
					EventDispatcher: makeEventDispatcher(),
				}),
				AddressFactory: fakeAddressFactory(),
			}).execute({
				eventRecorder: context.eventRecorder,
				state: context.state,
				tx: {
					data: {
						recipient: {
							toString: () => 'bear1vlezq7dtyml3zu3wgfdwdglcalradfnt5xpzahl7a5478hn5g20qxd5nkr',
							verify: () => true,
						},
					},
				},
			}),
		).resolves.not.toThrow()

		expect(
			await context.state
				.getAccountRepository()
				.hasByAddress('bear1vlezq7dtyml3zu3wgfdwdglcalradfnt5xpzahl7a5478hn5g20qxd5nkr'),
		).toBeTrue()
	})

	it('should not create a new account if any exceptions are thrown', async (context) => {
		context.eventRecorder.actingAs('find-or-create-recipient-does-not-exist')

		expect(
			await context.state
				.getAccountRepository()
				.hasByAddress('bear1vlezq7dtyml3zu3wgfdwdglcalradfnt5xpzahl7a5478hn5g20qxd5nkr'),
		).toBeFalse()

		await expect(
			makeFindOrCreateRecipientListener({
				AccountSerializer: makeAccountSerializer({
					CommittedState: context.state,
					DeliverTxState: context.state,
					EventDispatcher: makeEventDispatcher(),
				}),
				AddressFactory: () => {
					throw new Error('Hello')
				},
			}).execute({
				eventRecorder: context.eventRecorder,
				state: context.state,
				tx: {
					data: {
						recipient: {
							toString: () => 'bear1vlezq7dtyml3zu3wgfdwdglcalradfnt5xpzahl7a5478hn5g20qxd5nkr',
							verify: () => true,
						},
					},
				},
			}),
		).toThrowAttributes({
			attributes: [
				{
					index: false,
					key: 'recipient',
					value: 'ERR_TX_MALFORMED_ADDRESS',
				},
			],
		})

		expect(
			await context.state
				.getAccountRepository()
				.hasByAddress('bear1vlezq7dtyml3zu3wgfdwdglcalradfnt5xpzahl7a5478hn5g20qxd5nkr'),
		).toBeFalse()
	})
})
