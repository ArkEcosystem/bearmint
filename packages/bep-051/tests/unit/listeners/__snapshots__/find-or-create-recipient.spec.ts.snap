// Vitest Snapshot v1

exports[`Find or Create Recipient > should capture an event if the tx recipient cannot be verified 1`] = `
Event {
  "attributes": [],
  "type": "find-or-create-recipient-fail-verify",
}
`;

exports[`Find or Create Recipient > should exit early if the recipient account already exists 1`] = `
Event {
  "attributes": [],
  "type": "find-or-create-recipient-already-exists",
}
`;
